### DISCUSSION TITLE
[HW3] `Z3.Fixedpoint.query_r` returns `Z3.Solver.UNKNOWN` when there are multiple assertions

### DISCUSSION THREAD

**Discussion by ironore15** [prosyslab-classroom/cs424-program-reasoning#432] [Q&A] [answered]
### Your Name

민승기

### Question

`Z3.Fixedpoint.query_r`을 이용하여 assert가 두 개 이상인 smt를 풀려고 할 때, 항상 결과로 `Z3.Solver.UNKNOWN`이 나옵니다.

```c
#include "verifier.h"

int main() {
  int x = input();
  assume(x > 1);
  assert(x == 0);
  assert(x > 0);
  return 0;
}
```

```smt
(declare-rel if.then5 (Bool Bool Int Int Bool))
(declare-rel if.then (Bool Int Int Bool))
(declare-rel if.end7 (Bool Bool Int Int Bool))
(declare-rel entry ())
(declare-rel if.else6 (Bool Bool Int Int Bool))
(declare-rel if.else (Bool Int Int Bool))
(declare-rel if.end (Bool Int Int Bool))
(declare-var A Bool)
(declare-var B Int)
(declare-var C Int)
(declare-var D Bool)
(declare-var E Bool)
(rule (! entry :named entry!))
(rule (=> (and entry (= D (> C 1)) (= B (ite D 1 0)) (not (= 0 B)) (= A (= C 0)) A)
    (if.then A B C D)))
(rule (=> (and entry
         (= D (> C 1))
         (= B (ite D 1 0))
         (not (= 0 B))
         (= A (= C 0))
         (not A))
    (if.else A B C D)))
(rule (=> (and (if.then A B C D)) (if.end A B C D)))
(rule (=> (and (if.end D B C E) (= A (> C 0)) A) (if.then5 A D B C E)))
(rule (=> (and (if.end D B C E) (= A (> C 0)) (not A)) (if.else6 A D B C E)))
(rule (=> (and (if.then5 A D B C E)) (if.end7 A D B C E)))

(query if.else6)(query if.else)
```

해당 formula.smt2를 z3 명령어로 풀었을 때는 다음과 같이 잘 해결하지만
```
$ z3 formula.smt2            
unsat
sat
```

ocaml API의 `Z3.Fixedpoint.query_r`로 풀었을 경우에는 `Z3.Solver.UNKNOWN`이 리턴됩니다.
디버깅을 위해 `Z3.Fixedpoint.get_reason_unknown solver`를 확인해본 결과 `"ok"`로 나와 원인을 모르겠네요.

과제 스펙이 multiple assertion을 포함할 수 있나요? 만약 그런 경우, `Z3.Fixedpoint.query_r`을 위해 제가 추가적으로 더해줘야 하는 룰이 있나요?

---

When I attempt to solve an SMT problem with more than one assertion using `Z3.Fixedpoint.query_r`, I always receive the result `Z3.Solver.UNKNOWN`.

Here's the code example in C:
```c
#include "verifier.h"

int main() {
  int x = input();
  assume(x > 1);
  assert(x == 0);
  assert(x > 0);
  return 0;
}
```

And the SMT formula:
```smt
(declare-rel if.then5 (Bool Bool Int Int Bool))
(declare-rel if.then (Bool Int Int Bool))
(declare-rel if.end7 (Bool Bool Int Int Bool))
(declare-rel entry ())
(declare-rel if.else6 (Bool Bool Int Int Bool))
(declare-rel if.else (Bool Int Int Bool))
(declare-rel if.end (Bool Int Int Bool))
(declare-var A Bool)
(declare-var B Int)
(declare-var C Int)
(declare-var D Bool)
(declare-var E Bool)
(rule (! entry :named entry!))
(rule (=> (and entry (= D (> C 1)) (= B (ite D 1 0)) (not (= 0 B)) (= A (= C 0)) A)
    (if.then A B C D)))
(rule (=> (and entry
         (= D (> C 1))
         (= B (ite D 1 0))
         (not (= 0 B))
         (= A (= C 0))
         (not A))
    (if.else A B C D)))
(rule (=> (and (if.then A B C D)) (if.end A B C D)))
(rule (=> (and (if.end D B C E) (= A (> C 0)) A) (if.then5 A D B C E)))
(rule (=> (and (if.end D B C E) (= A (> C 0)) (not A)) (if.else6 A D B C E)))
(rule (=> (and (if.then5 A D B C E)) (if.end7 A D B C E)))

(query if.else6)(query if.else)
```

When I solve this formula.smt2 using the z3 command, it correctly gives the following result:
```
$ z3 formula.smt2            
unsat
sat
```

However, when using the OCaml API `Z3.Fixedpoint.query_r`, it returns `Z3.Solver.UNKNOWN`. I checked `Z3.Fixedpoint.get_reason_unknown solver` for debugging, but it only returns "ok", leaving me unsure of the cause.

Does the project specification include support for multiple assertions? If so, are there additional rules that I need to add for `Z3.Fixedpoint.query_r` to handle them?

**Comment by HyunjoonCho** [comment, marked as answer]
README의 Setup 단락에 따르면, assert문은 한 개만 존재한다고 합니다.
> Each input program will have one assert statement. The VC is satisfiable if and only if there exists a program execution that invalidates the assertion. The verifier checks whether the VC is satisfiable (i.e., assertion fail) or unsatisfiable (i.e., assertion verified). Initially, you will always see Verified since you have not implemented the VC generator part yet.

============

According to the Setup section of the README file, there is only one assert statement in the input program.
> Each input program will have one assert statement. The VC is satisfiable if and only if there exists a program execution that invalidates the assertion. The verifier checks whether the VC is satisfiable (i.e., assertion fail) or unsatisfiable (i.e., assertion verified). Initially, you will always see Verified since you have not implemented the VC generator part yet.

**Reply by ironore15** [reply]
답변 감사합니다. 혹시 Z3 api에서 해당 문제를 어떻게 해결하는지, 그리고 `Z3.Fixedpoint.query`와 `Z3.Fixedpoint.query_r`의 차이점도 아신다면 답변해주시면 감사하겠습니다. ocaml의 경우 z3 documentation이 좀 부족해서 찾기가 어렵네요.

---

Thank you for your response. If you know how to solve this issue with the Z3 API and understand the difference between `Z3.Fixedpoint.query` and `Z3.Fixedpoint.query_r`, I would appreciate your help. Documentation for Z3 in OCaml is somewhat lacking, making it hard to find information.

**Reply by spearo2** [reply]
[Z3 doc](https://z3prover.github.io/api/html/ml/Z3.Fixedpoint.html), 해당 자료를 확인하시면 도움 되실것 같습니다.

==========

[Z3 doc](https://z3prover.github.io/api/html/ml/Z3.Fixedpoint.html), Checking this document will help.
